pkg/artifact/location_data.go

```diff
@@ -0,0 +1,64 @@
+package artifact
+
+import (
+	"net/url"
+
+	"github.com/square/p2/pkg/auth"
+	"github.com/square/p2/pkg/types"
+	"github.com/square/p2/pkg/util"
+)
+
+// Given a launchable stanza, returns the URL from which the artifact may be downloaded, as
+// well as an auth.VerificationData which can be used to verify the artifact.
+// There are two schemes for specifying this information in a launchable stanza:
+// 1) using the "location" field. In this case, the artifact location is simply the value
+// of the field and the path to the verification files is inferred using magic suffixes
+// 2) the "version" field is provided. In this case, the artifact registry is queried with
+// the information specified under the "version" key and the response contains the URLs
+// from which the extra files may be fetched, and these are returned.
+//
+// When using the first method, the following magical suffixes are assumed:
+// manifest: ".manifest"
+// manifest signature: ".manifest.sig"
+// build signature: ".sig"
+func LocationDataForLaunchable(stanza types.LaunchableStanza) (*url.URL, auth.VerificationData, error) {
+	if stanza.Location == "" && stanza.Version.ID == "" {
+		return nil, auth.VerificationData{}, util.Errorf("Launchable must provide either \"location\" or \"version\" fields")
+	}
+
+	if stanza.Location != "" && stanza.Version.ID != "" {
+		return nil, auth.VerificationData{}, util.Errorf("Launchable must not provide both \"location\" and \"version\" fields")
+	}
+
+	// infer the verification data using magical suffixes
+	if stanza.Location != "" {
+		location, err := url.Parse(stanza.Location)
+		if err != nil {
+			return nil, auth.VerificationData{}, util.Errorf("Couldn't parse launchable url '%s': %s", stanza.Location, err)
+		}
+
+		verificationData := VerificationDataForLocation(location)
+		return location, verificationData, nil
+	}
+
+	return nil, auth.VerificationData{}, util.Errorf("Version key in launchable stanza not implemented")
```

Comment:
sure, i did make a PR comment saying this is the intended future but I suppose I can do that now even if it doesn't do anything interesting yet